src/libc/mod.rs

```diff
@@ -0,0 +1,323 @@
+                                 attr: *mut pthread_attr_t)
+    -> c_int
+{
+    pthread_attr_init(attr)
+}
+*/
+
+pub unsafe fn pthread_create(pthread: *mut pthread_t,
+                             attr: *const pthread_attr_t,
+                             start_routine: extern "C" fn(*mut c_void) -> *mut c_void,
+                             arg: *mut c_void)
+    -> c_int
+{
+    let _ = attr;
+    let flags = CLONE_VM | CLONE_FS | CLONE_FILES | CLONE_SIGHAND
+        | CLONE_THREAD | CLONE_SYSVSEM | CLONE_SETTLS
+        | CLONE_CHILD_CLEARTID | CLONE_PARENT_SETTID;
+
+    let align = 16;
+    let mask = align - 1;
+    let stack_size = ((*attr).stack_size + mask) & !mask;
+
+    let map = mmap(ptr::null_mut(),
+                   stack_size + mem::size_of::<thread>(),
+                   PROT_READ | PROT_WRITE,
+                   MAP_PRIVATE | MAP_ANON,
+                   -1,
+                   0);
+
+    // musl: src/internal/__syscall_ret.c
+    if map as usize > -4096isize as usize {
+        return -(map as c_int);
+    }
+
+    let stack = map.offset(stack_size as isize);
+    let thread = stack as *mut thread;
+    (*thread).this = thread;
+
+    let child_tid = syscall_clone(start_routine,
+                                  stack,
+                                  flags,
+                                  arg,
+                                  &mut (*thread).thread_id,
+                                  thread as *mut c_void,
+                                  &mut (*thread).thread_id);
+    if child_tid < 0 {
+        return -child_tid;
+    }
+    *pthread = pthread_t {
+        thread: thread,
+    };
+    0
+}
+
+extern {
+    // Defined in internal/<arch>.rs.
+    //
+    // Does the equivalent of
+    //
+    // ```
+    // let result = CLONE(flags, child_stack, ptid, ctid, newtls);
+    // if result == 0 {
+    //     EXIT(fn_(arg));
+    // } else {
+    //     result
+    // }
+    // ```
+    //
+    // where CLONE and EXIT are the respective syscalls. Because it mangles
+    // with the stack (in two processes at once), it needs to be written in the
+    // target's assembly.
+    #[link_name = "__steed_clone"]
+    fn syscall_clone(fn_: extern "C" fn(*mut c_void) -> *mut c_void,
+                     child_stack: *mut c_void,
+                     flags: c_ulong,
+                     arg: *mut c_void,
+                     ptid: *mut pid_t,
+                     newtls: *mut c_void,
+                     ctid: *mut pid_t) -> pid_t;
+}
+
+/*
+#[inline(always)]
+#[cfg(target_arch = "x86_64")]
+unsafe fn thread_self() -> *mut thread {
+    let result;
+    asm!("mov %fs:0,$0":"=r"(result));
+    result
+}
+
+#[inline(always)]
+pub unsafe fn pthread_self() -> pthread_t {
+    pthread_t {
+        thread: thread_self(),
+    }
+}
+
+pub unsafe fn pthread_detach(thread: pthread_t) -> c_int {
+    unimplemented!();
+}
+*/
+
+pub unsafe fn pthread_join(pthread: pthread_t, retval: *mut *mut c_void)
```

Comment:
This implementation looks different from the MUSL one, where did it come from? (I checked `src/thread/pthread_join.c`). Regardless, it looks reasonable to me.